synonyms/directionals/de.txt

```diff
@@ -0,0 +1,33 @@
+nord, n
+nördlich, nördl, nordl, nordlich, noerdlich
```

Comment:
aren't diacritics similar to punctuation in that they'll be stripped earlier in the analysis pipeline than synonym handling?

I suppose it won't make a difference, since we will remove duplicate tokens even later, right?

Reply:
Yeah I was actually going to leave a comment about that...

You're right, as-is this might produce some tokens which will never be matched, which isn't ideal since it will take up extra disk space, but won't do any harm.

There's a few reasons I chose to leave them in:
- the `icu_folding` filter implemention is currently inconsistent, in some places it's run before the synonyms and others after
- I would consider removing the `icu_folding` filter at a later date since it can cause issues in some cases (like the recent Hütten vs. Hutten issue)
- Generating these lists is super time-consuming and I'd prefer to not have to do it all again any time soon

Reply:
Yeah, that makes sense. We can afford the extra disk space in the short/medium term. It makes sense to not sweat the details while we test out these synonyms.

Here's an idea for the longer term: what if we made a synonyms library that had many lists of synonyms and could process and expose them as a Javascript module in many different ways. The processed output could automatically exclude synonyms that differ only by diacritical.

It would also be helpful since there are at least 3 very different places where we probably want many of the same synonyms:
- here, in pelias/schema for Elasticsearch synonyms. These will ensure we match properly even when input/data have different values
- In pelias/api's deduplication layer. This allows us to deduplicate records that have, for example, differing abbreviations but refer to the same place
- In pelias/fuzzy-tester, so that we can generate resilient test cases that don't have to care if their expectations exactly match what we happen to have in our data.

Reply:
I also like idea of running `brücke, brucke, bruecke` before the `icu_folding` filter because it produces 3 forms (umlaut, ASCII, English-keyboard) where we are only able to currently produce two of those.

Reply:
Yeah also `pelias/parser` and `libpostal` use these synonyms in some form so they're used throughout.

Reply:
For synonyms like `café, cafe` once the `icu_folding` filter runs it'll produce two tokens with the same value at the same position, there is a `unique_only_same_position` filter in place to take care of these dupes already, so hopefully there is little to no overhead.

Reply:
In `pelias/parser` we are using [`remove-accents`](https://github.com/pelias/parser/blob/master/tokenization/normalizer.js#L1) when we load dictionaries. Could that be used here too?

Reply:
I would prefer not to, the `icu_folding` filter ensures that only ASCII tokens are inserted in the index and the `unique_only_same_position` filter removes any duplicate tokens at the same position, so `café, cafe` will still only generate one token.

Reply:
There are benefits to having the synonyms come *before* the `icu_folding` filter, such as the `brücke, brucke, bruecke` case above.

My suspicion is that stripping the diacritics earlier in the pipeline will become an impediment to some languages in the future because there will be no way for users to generate tokens based on the native language representation of the word.

I'm not that well versed in languages but I suspect it's not a big problem in Western Europe but the further East you go the more valuable this would be to have more control over.

Reply:
One interesting thing I've found is that, regardless of whether the `icu_folding` filter comes before or after the synonyms, either way, no tokens with diacritics are produces as a result:

```
elyzer --index pelias --analyzer peliasPhrase 'east west café'
CHAR_FILTER: punctuation
{0:east west café}
CHAR_FILTER: nfkc_normalizer
{0:east west café}
TOKENIZER: peliasNameTokenizer
{0:east}	{1:west}	{2:café}
TOKEN_FILTER: lowercase
{0:east}	{1:west}	{2:café}
TOKEN_FILTER: trim
{0:east}	{1:west}	{2:café}
TOKEN_FILTER: remove_duplicate_spaces
{0:east}	{1:west}	{2:café}
TOKEN_FILTER: synonyms/punctuation
{0:east}	{1:west}	{2:café}
TOKEN_FILTER: synonyms/custom_name
{0:east}	{1:west}	{2:café}
TOKEN_FILTER: synonyms/personal_titles
{0:east}	{1:west}	{2:café}
TOKEN_FILTER: synonyms/place_names
{0:east}	{1:west}	{2:café,cafe}
TOKEN_FILTER: synonyms/streets
{0:east}	{1:west}	{2:café,cafe}
TOKEN_FILTER: synonyms/directionals
{0:east,e}	{1:west,w,wst}	{2:café,cafe}
TOKEN_FILTER: icu_folding
{0:east,e}	{1:west,w,wst}	{2:cafe,cafe}
TOKEN_FILTER: remove_ordinals
{0:east,e}	{1:west,w,wst}	{2:cafe,cafe}
TOKEN_FILTER: unique_only_same_position
{0:east,e}	{1:west,w,wst}	{2:cafe}
TOKEN_FILTER: notnull
{0:east,e}	{1:west,w,wst}	{2:cafe}
TOKEN_FILTER: flatten_graph
{0:east,e}	{1:west,w,wst}	{2:cafe}
```

```
elyzer --index pelias --analyzer peliasIndexOneEdgeGram 'east west café'
CHAR_FILTER: punctuation
{0:east west café}
CHAR_FILTER: nfkc_normalizer
{0:east west café}
TOKENIZER: peliasNameTokenizer
{0:east}	{1:west}	{2:café}
TOKEN_FILTER: lowercase
{0:east}	{1:west}	{2:café}
TOKEN_FILTER: icu_folding
{0:east}	{1:west}	{2:cafe}
TOKEN_FILTER: trim
{0:east}	{1:west}	{2:cafe}
TOKEN_FILTER: synonyms/custom_name
{0:east}	{1:west}	{2:cafe}
TOKEN_FILTER: synonyms/personal_titles
{0:east}	{1:west}	{2:cafe}
TOKEN_FILTER: synonyms/place_names
{0:east}	{1:west}	{2:cafe,cafe}
TOKEN_FILTER: synonyms/streets
{0:east}	{1:west}	{2:cafe,cafe}
TOKEN_FILTER: synonyms/directionals
{0:east,e}	{1:west,w,wst}	{2:cafe,cafe}
TOKEN_FILTER: synonyms/punctuation
{0:east,e}	{1:west,w,wst}	{2:cafe,cafe}
TOKEN_FILTER: remove_ordinals
{0:east,e}	{1:west,w,wst}	{2:cafe,cafe}
TOKEN_FILTER: removeAllZeroNumericPrefix
{0:east,e}	{1:west,w,wst}	{2:cafe,cafe}
TOKEN_FILTER: peliasOneEdgeGramFilter
{0:e,ea,eas,east,e}	{1:w,we,wes,west,w,w,ws,wst}	{2:c,ca,caf,cafe,c,ca,caf,cafe}
TOKEN_FILTER: unique_only_same_position
{0:e,ea,eas,east}	{1:w,we,wes,west,ws,wst}	{2:c,ca,caf,cafe}
TOKEN_FILTER: notnull
{0:e,ea,eas,east}	{1:w,we,wes,west,ws,wst}	{2:c,ca,caf,cafe}
TOKEN_FILTER: flatten_graph
{0:e,ea,eas,east}	{1:w,we,wes,west,ws,wst}	{2:c,ca,caf,cafe}
```

Reply:
Another interesting thing I found is that the synonyms don't 'explode' like I assumed they did.
(I was assuming that because `street, str` and `straße, str` that this meant `street, straße`, fortunately this isn't true!

This is great, so as long as the street names are indexed in their expanded form (ie. in the source data as 'Street', not 'St.') then they will only get the minimal amount of synonyms to be correct, whereas in the case where the source provides an abbreviated for which is ambiguous more synonyms are produced:

```
elyzer --index pelias --analyzer peliasStreet 'main str'
CHAR_FILTER: punctuation
{0:main str}
CHAR_FILTER: nfkc_normalizer
{0:main str}
TOKENIZER: peliasStreetTokenizer
{0:main}	{1:str}
TOKEN_FILTER: lowercase
{0:main}	{1:str}
TOKEN_FILTER: trim
{0:main}	{1:str}
TOKEN_FILTER: remove_duplicate_spaces
{0:main}	{1:str}
TOKEN_FILTER: synonyms/custom_street
{0:main}	{1:str}
TOKEN_FILTER: synonyms/personal_titles
{0:main}	{1:str}
TOKEN_FILTER: synonyms/streets
{0:main}	{1:str,straße,strasse,street,st,stre,stree,strt}
TOKEN_FILTER: synonyms/directionals
{0:main}	{1:str,straße,strasse,street,st,stre,stree,strt}
TOKEN_FILTER: icu_folding
{0:main}	{1:str,strasse,strasse,street,st,stre,stree,strt}
TOKEN_FILTER: remove_ordinals
{0:main}	{1:str,strasse,strasse,street,st,stre,stree,strt}
TOKEN_FILTER: trim
{0:main}	{1:str,strasse,strasse,street,st,stre,stree,strt}
TOKEN_FILTER: unique_only_same_position
{0:main}	{1:str,strasse,street,st,stre,stree,strt}
TOKEN_FILTER: notnull
{0:main}	{1:str,strasse,street,st,stre,stree,strt}
TOKEN_FILTER: flatten_graph
{0:main}	{1:str,strasse,street,st,stre,stree,strt}
```

```
elyzer --index pelias --analyzer peliasStreet 'main street'
CHAR_FILTER: punctuation
{0:main street}
CHAR_FILTER: nfkc_normalizer
{0:main street}
TOKENIZER: peliasStreetTokenizer
{0:main}	{1:street}
TOKEN_FILTER: lowercase
{0:main}	{1:street}
TOKEN_FILTER: trim
{0:main}	{1:street}
TOKEN_FILTER: remove_duplicate_spaces
{0:main}	{1:street}
TOKEN_FILTER: synonyms/custom_street
{0:main}	{1:street}
TOKEN_FILTER: synonyms/personal_titles
{0:main}	{1:street}
TOKEN_FILTER: synonyms/streets
{0:main}	{1:street,st,str,stre,stree,strt}
TOKEN_FILTER: synonyms/directionals
{0:main}	{1:street,st,str,stre,stree,strt}
TOKEN_FILTER: icu_folding
{0:main}	{1:street,st,str,stre,stree,strt}
TOKEN_FILTER: remove_ordinals
{0:main}	{1:street,st,str,stre,stree,strt}
TOKEN_FILTER: trim
{0:main}	{1:street,st,str,stre,stree,strt}
TOKEN_FILTER: unique_only_same_position
{0:main}	{1:street,st,str,stre,stree,strt}
TOKEN_FILTER: notnull
{0:main}	{1:street,st,str,stre,stree,strt}
TOKEN_FILTER: flatten_graph
{0:main}	{1:street,st,str,stre,stree,strt}
```

Reply:
That sounds great! :tada: